// thebe-core/dist/cjs/thebe/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupThebeCore = exports.setupNotebookFromIpynb = exports.setupNotebookFromBlocks = exports.makeServer = exports.makeEvents = exports.connectToJupyterLite = exports.connectToJupyter = exports.connectToBinder = void 0;
const tslib_1 = require("tslib");
const server_1 = tslib_1.__importDefault(require("../server"));
const notebook_1 = tslib_1.__importDefault(require("../notebook"));
const events_1 = require("../events");
const options_1 = require("../options");
const rendermime_1 = require("../rendermime");
const coreModule = tslib_1.__importStar(require("../index"));
const version_1 = tslib_1.__importDefault(require("../version"));
function connectToBinder(config) {
    const server = new server_1.default(config);
    console.debug(`thebe:api:connect binder 📡`, config.binder);
    server.connectToServerViaBinder();
    return server;
}
exports.connectToBinder = connectToBinder;
function connectToJupyter(config) {
    const server = new server_1.default(config);
    console.debug(`thebe:api:connect direct 🔌`, config.serverSettings);
    server.connectToJupyterServer();
    return server;
}
exports.connectToJupyter = connectToJupyter;
function connectToJupyterLite(config) {
    const server = new server_1.default(config);
    console.debug(`thebe:api:connect JupyterLite 🤘`);
    server.connectToJupyterLiteServer();
    return server;
}
exports.connectToJupyterLite = connectToJupyterLite;
function makeEvents() {
    return new events_1.ThebeEvents();
}
exports.makeEvents = makeEvents;
function makeServer(config) {
    return new server_1.default(config);
}
exports.makeServer = makeServer;
function setupNotebookFromBlocks(blocks, config, rendermime) {
    return notebook_1.default.fromCodeBlocks(blocks, config, rendermime);
}
exports.setupNotebookFromBlocks = setupNotebookFromBlocks;
function setupNotebookFromIpynb(ipynb, config, rendermime) {
    return notebook_1.default.fromIpynb(ipynb, config, rendermime);
}
exports.setupNotebookFromIpynb = setupNotebookFromIpynb;
function setupThebeCore() {
    var _a;
    console.debug(`thebe-core (v${version_1.default})`, { coreModule });
    window.thebeCore = Object.assign((_a = window.thebeCore) !== null && _a !== void 0 ? _a : {}, {
        module: coreModule,
        api: {
            makeConfiguration: options_1.makeConfiguration,
            makeEvents,
            makeServer,
            makeRenderMimeRegistry: rendermime_1.makeRenderMimeRegistry,
            connectToBinder,
            connectToJupyter,
            connectToJupyterLite,
            setupNotebookFromBlocks,
            setupNotebookFromIpynb,
        },
        version: version_1.default,
    });
}
exports.setupThebeCore = setupThebeCore;
