// thebe-core/dist/cjs/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ensureCoreOptions = exports.makeConfiguration = exports.makeMathjaxOptions = exports.makeServerSettings = exports.makeKernelOptions = exports.makeSavedSessionOptions = exports.makeBinderOptions = void 0;
const config_1 = require("./config");
const utils_1 = require("./utils");
function makeBinderOptions(opts) {
    return Object.assign({ repo: 'executablebooks/thebe-binder-base', ref: 'HEAD', binderUrl: 'https://mybinder.org', repoProvider: 'github' }, opts);
}
exports.makeBinderOptions = makeBinderOptions;
function makeSavedSessionOptions(opts) {
    return Object.assign({ enabled: true, maxAge: 86400, storagePrefix: 'thebe-binder' }, opts);
}
exports.makeSavedSessionOptions = makeSavedSessionOptions;
function makeKernelOptions(opts) {
    var _a, _b;
    return {
        path: (_a = opts.path) !== null && _a !== void 0 ? _a : '/',
        kernelName: (_b = opts.kernelName) !== null && _b !== void 0 ? _b : 'python',
    };
}
exports.makeKernelOptions = makeKernelOptions;
function makeServerSettings(settings) {
    var _a, _b;
    const baseUrl = (_a = settings.baseUrl) !== null && _a !== void 0 ? _a : 'http://localhost:8888';
    const wsUrl = (_b = settings.wsUrl) !== null && _b !== void 0 ? _b : baseUrl.replace(/^http/, 'ws');
    return Object.assign(Object.assign({ token: (0, utils_1.shortId)(), appendToken: true }, settings), { wsUrl,
        baseUrl });
}
exports.makeServerSettings = makeServerSettings;
function makeMathjaxOptions(opts) {
    return Object.assign({ mathjaxUrl: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js', mathjaxConfig: 'TeX-AMS_CHTML-full,Safe' }, opts);
}
exports.makeMathjaxOptions = makeMathjaxOptions;
function makeConfiguration(options, events) {
    return new config_1.Config(options, { events });
}
exports.makeConfiguration = makeConfiguration;
function ensureCoreOptions(options, events) {
    const config = new config_1.Config(options, { events });
    return Object.assign(Object.assign({}, config.base), { binderOptions: config.binder, savedSessionOptions: config.savedSessions, kernelOptions: config.kernels, serverSettings: config.serverSettings });
}
exports.ensureCoreOptions = ensureCoreOptions;
