// thebe-core/dist/esm/thebe/api.js
import ThebeServer from '../server';
import ThebeNotebook from '../notebook';
import { ThebeEvents } from '../events';
import { makeConfiguration } from '../options';
import { makeRenderMimeRegistry } from '../rendermime';
import * as coreModule from '../index';
import version from '../version';
export function connectToBinder(config) {
    const server = new ThebeServer(config);
    console.debug(`thebe:api:connect binder 📡`, config.binder);
    server.connectToServerViaBinder();
    return server;
}
export function connectToJupyter(config) {
    const server = new ThebeServer(config);
    console.debug(`thebe:api:connect direct 🔌`, config.serverSettings);
    server.connectToJupyterServer();
    return server;
}
export function connectToJupyterLite(config) {
    const server = new ThebeServer(config);
    console.debug(`thebe:api:connect JupyterLite 🤘`);
    server.connectToJupyterLiteServer();
    return server;
}
export function makeEvents() {
    return new ThebeEvents();
}
export function makeServer(config) {
    return new ThebeServer(config);
}
export function setupNotebookFromBlocks(blocks, config, rendermime) {
    return ThebeNotebook.fromCodeBlocks(blocks, config, rendermime);
}
export function setupNotebookFromIpynb(ipynb, config, rendermime) {
    return ThebeNotebook.fromIpynb(ipynb, config, rendermime);
}
export function setupThebeCore() {
    var _a;
    console.debug(`thebe-core (v${version})`, { coreModule });
    window.thebeCore = Object.assign((_a = window.thebeCore) !== null && _a !== void 0 ? _a : {}, {
        module: coreModule,
        api: {
            makeConfiguration,
            makeEvents,
            makeServer,
            makeRenderMimeRegistry,
            connectToBinder,
            connectToJupyter,
            connectToJupyterLite,
            setupNotebookFromBlocks,
            setupNotebookFromIpynb,
        },
        version,
    });
}
